tutorials/haskell/bdd-in-haskell-with-hspec/tutorial.md

```diff
@@ -0,0 +1,523 @@
+should end up looking like this:
+
+```
+library
+  hs-source-dirs:      src
+  exposed-modules:     Luhn
+                     , Luhn.Internal
+  build-depends:       base >= 4.7 && < 5
+  default-language:    Haskell2010
+```
+
+Lastly we'll add some shell content for the project to compile:
+
+```
+$ echo "module Luhn where" > src/Luhn.hs
+$ mkdir src/Luhn
+$ echo "module Luhn.Internal where" > src/Luhn/Internal.hs
+```
+
+Run the specs again and everything should be green, but we should
+have 0 examples now.
+
+### Building the foundation
+
+When programming the BDD way, you start writing the specs first. This is not only
+a good practice, but also helps reason about your code, as a series of individual
+decoupled units.
+
+For the Luhn algorithm, we'll need a `validate :: Integer -> Bool` function
+that receives a number and determines if it is valid or not. Based on the Luhn
+algorithm description, we can foresee the need of the following helper functions:
+
+* `toDigits :: Integer -> [Integer]` to convert an integer number to a
+list of digits.
+* `doubleEveryOther :: [Integer] -> [Integer]` to double every other digit
+starting from the right.
+* `sumDigits :: [Integer] -> Integer` to sum all digits, ensuring doubled
+digits sum its own digits
+first.
+
+As a good practice some developers split their code into two source files.
+One of the modules will contain the function we want exposed publicly to the
+whole project. The other one (commonly suffixed as `.Internal`) will hold
+the helper functions.
+
+You should avoid importing the Internal modules in your application,
+but having this file around is useful whenever you want to import them in order
+to test its functions.
+
+So we are going to edit our `src/Luhn.hs` module file and create a stub function.
+The file should look like this:
+
+```haskell
+module Luhn
+  ( validate
+  )
+  where
+
+import Luhn.Internal
+
+validate :: Integer -> Bool
+validate = undefined
+```
+
+Then we edit our `src/Luhn/Internal.hs` module file, and insert the following code:
+
+```haskell
+module Luhn.Internal
+  ( toDigits
+  , doubleEveryOther
+  , sumDigits
+  )
+  where
+
+toDigits :: Integer -> [Integer]
+toDigits = undefined
+
+doubleEveryOther :: [Integer] -> [Integer]
+doubleEveryOther = undefined
+
+sumDigits :: [Integer] -> Integer
+sumDigits = undefined
+```
+
+Even if we should start with the spec first, having this undefined definitions at
+hand will help us compile the spec and run them.
+
+### Our first spec
+Let's create the `test/LuhnSpec.hs` file inside the `test` folder. The name of the
+folder is important here. In order for Hspec to find our spec files
+you should follow certain rules:
+
+1. Spec files have to be placed into the same directory as the test
+driver `Spec.hs` file, or into a subdirectory.
+1. The name of a spec file has to end in Spec.hs; the module
+name has to match the file name.
+1. Each spec file has to export a top-level binding “spec” with type
+`Spec`.
```

Comment:
Yes this is a markdown convenience so that if you have large lists and add items in the middle you don't need to re-enumerate. Anyway, and because this lists is short I am using progressive numbers now.